sn32: shared matrix: switch to continuous pwm

dexter93 opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Switch from oneshot PWM to continuous. This will enable full LED output and minimize any flickering.
Key matrix scan rate is also heavily improved from ~60ms avg down to ~25ms avg with lows at ~8ms
## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [ ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

dexter93 committed:
sn32: shared matrix: switch to continuous pwm

dexter93 committed:
248b keyboards: use default rgb configs

dexter93 committed:
248b: rgb: speed improvements

dexter93 committed:
use the row offset as channel indicator

dexter93 committed:
don't disable channels

dexter93 committed:
let's break it

dexter93 committed:
minify

dexter93 committed:
un-break it

dexter93 committed:
only interact with led matrix channels

dexter93 committed:
Revert "only interact with led matrix channels"

This reverts commit 6dac6d74e59e8e3fae5b47ae03b7a4bf13685d98.

dexter93 committed:
check the config

dexter93 committed:
reintroduce checks

dexter93 committed:
weird race condition but ok

dexter93 committed:
buy some time

dexter93 committed:
don't lock in a scan line

dexter93 committed:
test: run everything in a lock state

dexter93 committed:
test: Revert "reintroduce checks"

This reverts commit 0abc259ef4e265b150f3ffe33ec531579283846e.

dexter93 committed:
led rows low at init

dexter93 committed:
only count active leds in the matrix

dexter93 committed:
don't mess with the index

dexter93 committed:
direct control approach

dexter93 committed:
fix build

dexter93 committed:
simplify

dexter93 committed:
indicator switcheroo

dexter93 committed:
test: channel swap

dexter93 committed:
Revert "test: channel swap"

This reverts commit 9557bd27084f0986e5f8bb30266abe332f62161b.

dexter93 committed:
check if leds exist

dexter93 committed:
math?

dexter93 committed:
meth?

dexter93 committed:
slow down

dexter93 committed:
test: speeed

dexter93 committed:
Revert "test: speeed"

This reverts commit fcfe87c78a7987d2e5155b44196485e1e8746af6.

dexter93 committed:
Revert "slow down"

This reverts commit c22e62926b6b4cabddf2b954041773fffbd6915f.

dexter93 committed:
some matrix updates

dexter93 committed:
Revert "some matrix updates"

This reverts commit ce111d0a02c450f82ede0e13cd8a6bba58cd7fec.

dexter93 committed:
change current_row check
less math, more simple

dexter93 committed:
use the array size instead of the led size for updating

dexter93 committed:
limit lock to the pwm selection

IsaacDynamo commented on drivers/sn32/rgb_matrix_sn32f248b.c:
```diff
@@ -33,11 +33,9 @@
-static uint16_t row_ofsts[LED_MATRIX_ROWS];
 static uint8_t mr_offset[24] = {0};
 static uint8_t current_row = 0;
 LED_TYPE led_state[LED_MATRIX_ROWS * LED_MATRIX_COLS];
```
`LED_MATRIX_ROWS * LED_MATRIX_COLS` should be changed to `DRIVER_LED_TOTAL`

IsaacDynamo commented on drivers/sn32/rgb_matrix_sn32f248b.c:
```diff
@@ -355,15 +324,15 @@ void SN32F24XX_init(void) {
 
 void SN32F24XX_set_color_all(uint8_t r, uint8_t g, uint8_t b) {
-    for (int i=0; i<DRIVER_LED_TOTAL; i++)
+    for (int i = 0; i < sizeof(led_state) / sizeof(led_state[0]); i++) {
```
`sizeof(led_state) / sizeof(led_state[0])` can be changed back to `DRIVER_LED_TOTAL` when `led_state` is made `DRIVER_LED_TOTAL` in size.

IsaacDynamo commented on drivers/sn32/rgb_matrix_sn32f248b.c:
```diff
@@ -355,15 +324,15 @@ void SN32F24XX_init(void) {
 static void flush(void) {}
 
 void SN32F24XX_set_color(int index, uint8_t r, uint8_t g, uint8_t b) {
```
Functions SN32F24XX_init(), SN32F24XX_set_color(), SN32F24XX_set_color_all() can be made static.
And should be removed from rgb_matrix_sn32f24xx.h
The SN32F24XX_ prefix is also no longer needed, and can be removed.

IsaacDynamo commented on drivers/sn32/rgb_matrix_sn32f248b.c:
```diff
@@ -267,86 +265,57 @@ void shared_matrix_rgb_disable(void) {
         pwmDisableChannel(&PWMD1,i);
         chSysUnlockFromISR();
```
Why are the chSysLockFromISR() and chSysUnlockFromISR() calls needed here?

dexter93 committed:
rgb: get rid of phantom leds

dexter93 committed:
code improvements: speed

dexter93 committed:
testing: high on init

dexter93 committed:
gpio tidyup

dexter93 committed:
testing: priority

dexter93 committed:
more priority tests

dexter93 committed:
move the check in the interrupt

dexter93 committed:
dynamically adjust scan freq

dexter93 committed:
typefix

dexter93 committed:
testing: dead time

dexter93 committed:
only disable leds in the shared matrix

dexter93 committed:
check status before updating the buffer

dexter93 committed:
0 is a valid value

dexter93 committed:
testing again

dexter93 committed:
disable pwm only when we need to scan keys

dexter93 committed:
Merge remote-tracking branch 'origin/sn32' into pwm_cont

dexter93 committed:
gpio work only on scan

dexter93 committed:
Revert "gpio work only on scan"

This reverts commit d0f48bf0b25169998fb841c5201d2eaae9b25da8.

dexter93 committed:
first row testing

dexter93 committed:
Revert "first row testing"

This reverts commit dafb28714f529a35ce3b83190c321132d125713e.

dexter93 committed:
testing: keep the interrupt going

dexter93 committed:
Revert "testing: keep the interrupt going"

This reverts commit 6d5dbc1929be3eb568cc7e8f77cd0edb59f2a8c1.

dexter93 committed:
basic flush support

dexter93 committed:
Revert "basic flush support"

This reverts commit 4723c83999c0489cc05bec423b29d4cc60cff0ae.

dexter93 committed:
testing: freq

dexter93 committed:
this is a hack

dexter93 committed:
Declare our byte order
This reverts commit cb1469efad72183634f531fa335b7c3a82422ed2.

dexter93 committed:
guard the led index

dexter93 committed:
less hardcoding of indicators

dexter93 committed:
testing: advance last

dexter93 committed:
Revert "testing: advance last"

This reverts commit 9df003713d04d2955b3c6703e58146b3c8c0c95a.

dexter93 committed:
testing: disable rows too

dexter93 committed:
buffer check before updating

dexter93 committed:
kill it

dexter93 committed:
testing: index

dexter93 committed:
Revert "testing: index"

This reverts commit 55524788a5b66e698d8c645c5bef6ca38aaa7c87.

dexter93 committed:
testing: simple counter advance

dexter93 committed:
testing: the match enables the leds

dexter93 committed:
Revert "testing: simple counter advance"

This reverts commit d5440291564f21f3a2bff11817dfa96a6e613d53.

dexter93 committed:
killem'all

dexter93 committed:
testing this again

dexter93 committed:
index test

dexter93 committed:
more testing here

dexter93 committed:
Revert "more testing here"

This reverts commit 25f9cb59b4ef0c1943068d0b9ab7d2ae8e65d3f6.

dexter93 committed:
Revert "testing this again"

This reverts commit bd4f7fb15c97b4c96ec4d39e4d340f8fd6370bc6.

dexter93 committed:
remove led checks, change row id logic

dexter93 committed:
advance on last match

dexter93 committed:
align it

dexter93 committed:
advance per row, not col scan

dexter93 committed:
testing init at 0

dexter93 committed:
Revert "testing init at 0"

This reverts commit e7a93fa36455bc39a6c915d8709e9415146fec51.

dexter93 committed:
channel switcheroo

dexter93 committed:
inherit from color.h

dexter93 committed:
Revert "channel switcheroo"

This reverts commit 2c8c53ee149b19722d9136172018a411f4913ed5.

dexter93 committed:
frequency adjust

dexter93 committed:
some tlc

dexter93 committed:
more cleanup

dexter93 committed:
once again

dexter93 committed:
build it

dexter93 committed:
add value defaults here

dexter93 committed:
we don't need this anymore

dexter93 committed:
don't kill the rows

dexter93 committed:
Revert "don't kill the rows"
better keep this. doesn't add flickering
This reverts commit d1ffd53a28b92bbb5bbff8bf63725f70617dc641.

dexter93 committed:
Merge branch 'sn32' into pwm_cont

dexter93 commented:
Col2Row is good to go.

dexter93 committed:
Merge branch 'sn32' into pwm_cont

dexter93 committed:
bug hunting

dexter93 committed:
testing

dexter93 committed:
last row

dexter93 committed:
bump row id at r0, not last

dexter93 committed:
this is annoying

dexter93 committed:
move the gpio handling inside the pwm update

dexter93 committed:
low

dexter93 committed:
switch last

dexter93 committed:
kill the last row on scan

dexter93 committed:
index update

dexter93 committed:
index again

dexter93 committed:
align please

dexter93 committed:
testing r5

dexter93 committed:
align the channels too

dexter93 committed:
Revert "align the channels too"

This reverts commit abee8c4b914bb90b2767922e8125ce91991b9f2d.

dexter93 committed:
last row of the led channels should advance the next row id

dexter93 committed:
nuke pwm

dexter93 committed:
this check worked. don't mess with it

dexter93 committed:
Revert "nuke pwm"
of course it won't show up now
This reverts commit 7464cdad328ab8c56f684e71095ba2ec2bef9f74.

dexter93 committed:
revert to original config

dexter93 committed:
let's break it

dexter93 committed:
Revert "let's break it"

This reverts commit 92e5f26948ffe3b681919e93ed00cbf66ce831d0.

dexter93 committed:
break it more

dexter93 committed:
low

dexter93 committed:
I pity anyone following this train

dexter93 committed:
back to normality

dexter93 committed:
only enable if we have something to write

dexter93 committed:
switch matrix scanning logic.

Instead of scanning the entire key matrix every 3rd row, scan each row or col on every cycle

dexter93 committed:
build it

dexter93 committed:
I said build it

dexter93 committed:
Revert "only enable if we have something to write"

This reverts commit ff3302b35b851b68994e0923f450ac3f140befda.

dexter93 committed:
testing: rgb bleeding through

dexter93 committed:
Merge remote-tracking branch 'origin/sn32' into pwm_cont

dexter93 committed:
add akko

dexter93 committed:
testing: status check

dexter93 committed:
make sure leds are high before scan

dexter93 committed:
move the check to pwm instead of gpio

all this for an akko bug

dexter93 committed:
Revert "move the check to pwm instead of gpio"

This reverts commit c76152e4f078af9bad82ca8272f5fd4b6dd6aa96.

dexter93 committed:
why you akko do that

dexter93 committed:
speedup building

dexter93 committed:
akko bug is akko
tagged as #wontfix. maybe another day

dexter93 committed:
do it before the scan

dexter93 committed:
build all keyboards
This reverts commit 9bda95de8fb1a7483db2cf68985b34ca4acc1034.

dexter93 committed:
rgb: minimize flickering

dexter93 committed:
Merge remote-tracking branch 'origin/sn32' into pwm_cont

dexter93 committed:
build k14 too

dexter93 committed:
use a flag to do a status check

dexter93 committed:
use the status check for scan too

dexter93 committed:
we only care about zero values

dexter93 committed:
limit the check to gpio, kill pwm always

dexter93 committed:
fix row2col

dexter93 committed:
gpio: limit selection on scan time

dexter93 commented:
Current shared matrix logic can be visualized here. https://tinyurl.com/y57nbk8s

dexter93 committed:
cols look like they need the high-z

dexter93 committed:
speed up building

dexter93 committed:
build it

dexter93 committed:
simplify

dexter93 committed:
remove pwm interference

dexter93 committed:
all the cols

dexter93 committed:
less flicker

dexter93 committed:
fix indexing

dexter93 committed:
match it

dexter93 committed:
Revert "cols look like they need the high-z"
led leakage is higher with this
This reverts commit 2ed5cc2076b779d3a08fc47ffcf169f8ca25dbe4.

dexter93 committed:
code cleanup

dexter93 committed:
build everything

This reverts commit ebc8554087f739228835d3e734e60213fecb17ae.

dexter93 committed:
row2col: scan b side too

dexter93 committed:
row2col: less flickering

dexter93 committed:
only track the used channels

dexter93 committed:
attempt to match it

dexter93 committed:
dirty way

dexter93 committed:
git dirty

dexter93 committed:
simplyf-i

dexter93 committed:
pick a col will you

dexter93 committed:
cleanup time

dexter93 committed:
Remove leftovers from private branch
This reverts commit 82bb7d2eba782f17d98b3e2f1292d3efdbe6460a.

dexter93 commented:
PR updated. Col2Row and Row2Col work properly now

dexter93 commented:
Reports say this takes care of https://github.com/SonixQMK/qmk_firmware/issues/157 too

dexter93 committed:
Documentation update
